server/index.js

```diff
@@ -135,14 +129,121 @@ var infer = function(req, res) {
     });
+}
+
+var infer = function(req, res) {
+	if(req.query.format && req.query.format == "image") {
+		return res.json({
+			error: "The inference server does not yet support returning a visualization of the prediction."
+		});
+	}
+
+    roboflow.tf.tidy(() => {
+
+        var start = Date.now();
+
+        var buffer = Buffer.from(req.body, "base64");
+        var arr = new Uint8Array(buffer);
+
+        var tensor;
+        try {
+            // works for decoding 3 channel (rgb) and 4 channel (rgba) images
+            tensor = roboflow.tf.node.decodeImage(arr, 3);
+        } catch(e) {
+            // handle 1 channel (grayscale) images
+            // which throw an error if you try to force them to be decoded with 3 channels
+            var channel = roboflow.tf.node.decodeImage(arr, 1);
+            tensor = roboflow.tf.stack([
+                channel,
+                channel,
+                channel
+            ], 2).squeeze();
+            roboflow.tf.dispose(channel);
+        }
+
+        var configuration = {
+            max_objects: Number.MAX_SAFE_INTEGER,
+            overlap: 0.3,
+            nms_threshold: 0.3,
+            threshold: 0.4,
+            tile: false,
+            tile_rows: 1,
+            tile_cols: 1,
+        };
+
+        if(req.query.overlap) {
+            req.query.overlap = parseFloat(req.query.overlap);
+            if(req.query.overlap > 1) req.query.overlap /= 100;
+            configuration.overlap = req.query.overlap;
+            configuration.nms_threshold = req.query.overlap;
+        }
+
+        if(req.query.confidence) {
+            req.query.confidence = parseFloat(req.query.confidence);
+            if(req.query.confidence > 1) req.query.confidence /= 100;
+            configuration.threshold = req.query.confidence;
+        }
+
+        //takes a tile query parameter like so tile=3x5
+        if(req.query.tile) {
+            if(false){
+                return res.json({
+                    error: "The inference server does not support this query parameter without a hosted image."
+                });
+            }
+            configuration.tile = true;
+
+            //ensure the tile query param looks like tile=300
+            if (!req.query.tile.match(/([0-9]+)/)){
+                return res.json({
+                    error: "Tile query parameter improperly formatted."
+                });
+            }
+            configuration.tile_rows = req.query.tile;
+            configuration.tile_cols = req.query.tile;
+        }
+
+
+        req.model.configure(configuration);
+
+        if(configuration.tile === true){
+            const tileDimensions = [configuration.tile_rows, configuration.tile_cols];
+            const paddedImage = padImage(tensor, tileDimensions);
+            const tiles = splitImage(paddedImage, tileDimensions);
+            var combinedResult = [];
+            async.eachOf(roboflow.tf.unstack(tiles), function(tile, index, cb) {
```

Comment:
Does `eachOf` vs `eachOfSeries` or `eachOfLimit` produce better performance? Or are they all roughly equivalent? I'm wondering if there's a risk that spamming the event loop (and then the gpu) with several hundred callbacks in parallel might cause a bottleneck (or potentially make the GPU run out of memory) but maybe TFjs is handling that orchestration in the backend for us & this lets the GPU and CPU bound operations run independently.

Reply:
Performance is roughly equivalent which was initially weird to me. I would have assumed firing all of them off at once would run the inferences in parallel but it doesn't seem to. I honestly don't think it's hitting the gpu any harder than normal because of this.